netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersDecoder.java

```diff
@@ -0,0 +1,498 @@
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.grpc.netty;
+
+import static io.grpc.internal.TransportFrameUtil.toRawSeralizedHeader;
+import static io.grpc.netty.Utils.TE_HEADER;
+import static io.netty.handler.codec.http2.Http2CodecUtil.DEFAULT_HEADER_TABLE_SIZE;
+import static io.netty.handler.codec.http2.Http2Error.COMPRESSION_ERROR;
+import static io.netty.handler.codec.http2.Http2Error.ENHANCE_YOUR_CALM;
+import static io.netty.handler.codec.http2.Http2Error.PROTOCOL_ERROR;
+import static io.netty.handler.codec.http2.Http2Exception.connectionError;
+import static io.netty.util.AsciiString.isUpperCase;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.handler.codec.http2.DefaultHttp2HeadersDecoder;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2HeaderTable;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.handler.codec.http2.Http2HeadersDecoder;
+import io.netty.handler.codec.http2.internal.hpack.Decoder;
+import io.netty.util.AsciiString;
+import io.netty.util.internal.PlatformDependent;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A {@link Http2HeadersDecoder} that allows to use custom {@link Http2Headers} implementations.
+ *
+ * <p>Some of the code is copied from Netty's {@link DefaultHttp2HeadersDecoder}.
+ */
+abstract class GrpcHttp2HeadersDecoder implements Http2HeadersDecoder,
+    Http2HeadersDecoder.Configuration {
+
+  private static final float HEADERS_COUNT_WEIGHT_NEW = 1 / 5f;
+  private static final float HEADERS_COUNT_WEIGHT_HISTORICAL = 1 - HEADERS_COUNT_WEIGHT_NEW;
+
+  private final int maxHeaderSize;
+
+  private final Decoder decoder;
+  private final Http2HeaderTable headerTable;
+  private float numHeadersGuess = 4;
+
+  GrpcHttp2HeadersDecoder(int maxHeaderSize) {
+    this.maxHeaderSize = maxHeaderSize;
+    decoder = new Decoder(maxHeaderSize, DEFAULT_HEADER_TABLE_SIZE, 32);
+    headerTable = new GrpcHttp2HeaderTable();
+  }
+
+  @Override
+  public Http2HeaderTable headerTable() {
+    return headerTable;
+  }
+
+  @Override
+  public int maxHeaderSize() {
+    return maxHeaderSize;
+  }
+
+  @Override
+  public Http2Headers decodeHeaders(ByteBuf headerBlock) throws Http2Exception {
+    try {
+      final GrpcHttp2InboundHeaders headers = newHeaders((int) numHeadersGuess);
+      decoder.decode(headerBlock, headers);
+      if (decoder.endHeaderBlock()) {
+        maxHeaderSizeExceeded();
+      }
+
+      numHeadersGuess = HEADERS_COUNT_WEIGHT_NEW * headers.numHeaders()
```

Comment:
@carl-mastrangelo 

I copied this code from Netty's `DefaultHttp2HeadersDecoder`.

Regarding the cost of floating point multiplication, this will be lost in the noise really. Decoding headers takes microseconds, and floating point multiplication takes 2 or 3 clock cycles. So I wouldn't worry about that.

I think we should have some kind of estimation logic. I am not sure if that's the right one. Something that tends to overestimate the size would be nice. Obviously, in the proxy scenario it's going to be difficult to have a good estimate.